### ISSUE TITLE
mv doesn't propagate gradients to matrix term

### ISSUE THREAD

**Issue by zkolter** [pytorch/pytorch#3303]
It seems like autodiff of matrix terms of .mv() doesn't behave as I'd expect.

```python
A = Variable(torch.randn(3,2), requires_grad=True)
x = Variable(torch.rand(2), requires_grad=True)
(A.mv(x)).sum().backward()
print(A.grad)
print(x.grad)

# Variable containing:
#  0  0
#  0  0
#  0  0
# [torch.FloatTensor of size 3x2]
# Variable containing:
# -0.6709
#  0.9995
# [torch.FloatTensor of size 2]
```

But if we use .mm(), things work fine:
```python
A = Variable(torch.randn(3,2), requires_grad=True)
x = Variable(torch.rand(2), requires_grad=True)
(A.mm(x[:,None])).sum().backward()
print(A.grad)
print(x.grad)

# Variable containing:
#  0.7123  0.2487
#  0.7123  0.2487
#  0.7123  0.2487
# [torch.FloatTensor of size 3x2]
# Variable containing:
# -0.6709
#  0.9995
# [torch.FloatTensor of size 2]

**Comment by albanD** [comment]
Hi,

As a temporary fix, you can use `(A.mv(x) * 1).sum().backward()`.

The problem is the behavior between the `sum` backward and the `mv` backward.
The sum returns a `0` strided gradient which is not supported by MKL/BLAS implementations of `ger` (used in the `mv` backward).
@apaszke not sure where this was introduced and how to fix this properly. Should the blas wrapper clone the input if it is `0` strided? Or should it just raise an error (meaning that we need to add some `clone`s in some places in the code like the sum backward)?

**Comment by apaszke** [comment]
It should clone the input. Still, this problem should have been caught. We must be missing some return code checks from BLAS

**Comment by apaszke** [comment]
Ah apparently these functions have no return codes and the docs clearly state that vector strides can't be 0... We have to fix those bindings
